src/socket.c

```diff
@@ -117,7 +117,7 @@ static int connSocketConnect(connection *conn,
     conn->state = CONN_STATE_CONNECTING;
 
     conn->conn_handler = connect_handler;
-    aeCreateFileEvent(server.el, conn->fd, AE_WRITABLE, conn->type->ae_handler, conn);
+    serverCreateFileEvent(conn->fd, AE_WRITABLE, conn->type->ae_handler, conn);
```

Comment:
Now the connection abstraction depend on server.h and its globals. It is a circular dependency.

Can we store the mutex in the aeEventLoop struct and let ae handle the mutex? If we want it to be optional, we can enable it with a special aeEnableMutex which sets a flag that the mutex shall be used. When it's enabled, the aeCreateFileEvent and other functions use the mutex automatically.

It seems like a good idea to me, but I'm not really sure, so please explain to me if I'm wrong.

Related, I see https://github.com/valkey-io/valkey/pull/477 adds a new connection. It's using aeCreateFileEvent too and there's a risk we forget about the mutex in that implementation. So, it's better if ae can use the mutex automatically for all connection types in the event loop.

Reply:
Not sure I understand what you mean by circular dependency. `connection.h`  doesn't include `server.h.` `connection.h` includes ae.h but with the `AE_EXCLUDE_EVENT_MODIFICATION` flag set. Hence, the new connection won't compile when trying to access `aeCreateFileEvent`. I tried to take the mutex to the `eventloop` struct, but it is more complex and less clean in my opinion. However, if you feel strongly about this, I will change it.

Reply:
Hm... Let me think about it. I want to make a logical decision, not one based on feelings. :)

Reply:
I've modified the PR to use two separate header files:

1. `ae_apidata.h`: This will include only the functions that modify the apidata state, which we want to guard.
2. `ae_eventloop.h`: This will contain all the remaining functions.

The `ae.h` file will still include both of these headers, and `ae.c` will continue to implement both sets of functions.

In `connection.h`, we will include only `ae_eventloop.h` to prevent direct access to the modification functions.
Finally, in `server.c`, we will include `ae.h` to provide access to all necessary functions.

Reply:
I also have the same "feeling" as Viktor. encapsulating the mutex inside the EventLoop sounds correct and I am not sure I understand the complexity - is it related to propagating the io_poll_state to the eventloop? I think in general this is probably also an EventLoop ownership right?

Reply:
I moved the mutex inside the eventloop struct. This required some changes in `ae.c`